### PR TITLE
Added wrapper types and Codable support

### PR DESCRIPTION
**Pull request by colemancda**
- Added wrapper types for JS APIs (`JSType`)
- Added [SwiftFoundation](https://github.com/PureSwift/SwiftFoundation) support for porting Foundation code
- Added `JSEncoder` for Codable support
- Added Promise support (`JSPromise`)
- Added [Web Bluetooth API](https://developers.google.com/web/updates/2015/07/interact-with-ble-devices-on-the-web) support (tested on Chrome)
- Added support for `console.log` and assertions

### CODE REVIEW

**Comment by colemancda** [comment]
Screen Shot 2020-06-05 at 08 08 16
Screen Shot 2020-06-05 at 08 14 57

**Comment by colemancda** [comment]
Screen Shot 2020-06-05 at 08 26 44

File: Sources/JavaScriptKit/JS Types/JSPromise.swift [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+//
+//  Created by Alsey Coleman Miller on 6/3/20.
+//
+
+/**
+ JavaScript Promise
+ 
+ A Promise is a proxy for a value not necessarily known when the promise is created. It allows you to associate handlers with an asynchronous action's eventual success value or failure reason. This lets asynchronous methods return values like synchronous methods: instead of immediately returning the final value, the asynchronous method returns a promise to supply the value at some point in the future.
+ */
+public final class JSPromise<Success>: JSType where Success: JSValueConvertible, Success: JSValueConstructible {
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JS Types/JSPromise.swift:13]: I think generic parameter `Success` doesn't match JavaScript's Promise use.

JavaScript's Promise allows users to return any type of values, so the requirement of single static type in JSPromise is too hard constraint.

File: Sources/JavaScriptKit/JS Types/JSType.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    
+    init?(_ jsObject: JSObjectRef)
+    
+    var jsObject: JSObjectRef { get }
+}
+
+internal extension JSType {
+    
+    func toString() -> String? {
+        return jsObject.toString.function?.apply(this: jsObject).string
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JS Types/JSType.swift:19]:
```suggestion
    func toString() -> String {
        return jsObject.toString.function.assert().apply(this: jsObject).string
```

File: Sources/JavaScriptKit/JSValueConvertible.swift [hunk 3]
```diff
@@ -57,69 +69,63 @@ extension JSObjectRef: JSValueConvertible {
...
+extension Dictionary: JSValueConvertible where Value == JSValueConvertible, Key == String {
     public func jsValue() -> JSValue {
-        Swift.Array<JSValueConvertible>.jsValue(self)()
+        let object = JSObject()
+        self.forEach { object[$0.key] = $0.value.jsValue() }
+        return .object(object.jsObject)
     }
 }
 
-extension Array: JSValueConvertible where Element == JSValueConvertible {
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JSValueConvertible.swift:86]: Please keep this extension for Element == JSValueConvertible to be available for existential

File: Sources/JavaScriptKit/JSValueConvertible.swift [hunk 4]
```diff
@@ -192,3 +198,15 @@ extension Array where Element: JSValueConvertible {
         Swift.Array<JSValueConvertible>.withRawJSValues(self)(body)
     }
 }
+
+extension JSValueConvertible where Self: Encodable {
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JSValueConvertible.swift:202]: These kinds of internal implementation shouldn't output log without explicit user configuration and I think JSValueConvertiable shouldn't be failable.

**Pull request closed without merging**